Added functionality to capture response from vault generic secret resources.

jcfergus opened the pull request:
This adds a "capture_response" setting to vault_generic_secret resources, which allows data that is only returned in a response to a Create/Write call to be accessed in the resultant data_json and data attributes.  (i.e. CSRs returned from pki endpoints and certificates returned from signing requests). This should provide a workable solution for #67 and #111  .  

[Note that this is my first foray into Go programming and has only been lightly tested.].

seanschade commented on vault/resource_generic_secret.go:
```diff
@@ -104,14 +118,50 @@ func genericSecretResourceWrite(d *schema.ResourceData, meta interface{}) error
+			return nil
+		}
+		jsonDataBytes, err := json.Marshal(secret.Data)
+		log.Printf("[DEBUG] Secret is %s", jsonDataBytes)
```
I don't know if you want to log this even under `DEBUG`.

jcfergus replied:
Oops, meant to remove that.  :-O  Will take care of that.

seanschade commented on vault/resource_generic_secret.go:
```diff
@@ -104,14 +118,50 @@ func genericSecretResourceWrite(d *schema.ResourceData, meta interface{}) error
+		for k, v := range secret.Data {
+			if vs, ok := v.(string); ok {
+				dataMap[k] = vs
+			} else {
```
You can get rid of the `else` block if you put a `continue` in the `if` block.

```
if vs, ok := v.(string); ok {
	dataMap[k] = vs
        continue
}
// Again ignoring error because we know this value
// came from JSON in the first place and so must be valid.
vBytes, _ := json.Marshal(v)
dataMap[k] = string(vBytes)
```

seanschade commented on vault/resource_generic_secret.go:
```diff
@@ -104,14 +118,50 @@ func genericSecretResourceWrite(d *schema.ResourceData, meta interface{}) error
+
+		d.Set("path", path)
+		return nil
+	} else {
```
You don't need this `else` either since the `if` block above returns.

jcfergus commented:
@seanschade Thanks!  Addressed those comments.  

@rorybrowne I haven't had a chance to look at your solution yet; still bogged down in the project that I need the functionality for.  I looked at the tests, and couldn't find any good examples that would be applicable to this particular use case.   I will take a look and revisit those in a bit.  

As far as the lease details, that makes sense - the case I was interested initially the lease information was irrelevant and I was trying to keep it simple.  I'll see about adding that or merging in what you've done.

stefan-caraiman commented:
Any updates regarding this one guys?

tyrannosaurus-becks commented on vault/resource_generic_secret.go:
```diff
@@ -104,13 +118,48 @@ func genericSecretResourceWrite(d *schema.ResourceData, meta interface{}) error
+		}
+		jsonDataBytes, err := json.Marshal(secret.Data)
+		if err != nil {
+			return fmt.Errorf("Error marshaling JSON for %q: %s", path, err)
```
Could this error start with a lower-case character instead?

tyrannosaurus-becks commented:
@jcfergus hi! Seems like this PR is inactive, so I'm closing for now. If you get a chance to circle back to it, feel free to reopen it. Thank you so much!

pschulten commented:
---
**NOTE TO FUTURE SELF**

( when I'm here)

* do not do this
* always create certificates during runtime (with sidecars, initContainers, ...)
* make them short-lived

---
